### PR TITLE
dev/report#53: search on relationship and case (2)

### PR DESCRIPTION
**Pull request by jaapjansma**
Overview
----------------------------------------
When doing an advanced search with case parameters set and displaying related contacts. Gives all clients of all the found cases which have this relationship. What we would expect is that related contact is linked to the found cases. 

Steps to reproduce
--------------------------

Preparation:
1. Create a contact **Contact A**
2. Create another contact **Contact B**
3. Create a third contact **Contact C**
4. Create a case of type **Housing Support** for **Contact A**
5. On the case assign the role **Benefit specialist is** to **Contact B**
6. Create a second case of type **Adult day care referral**
7. On the case assign the role **Benefit specialist is** to **Contact C**

Searching:
1. Go to advanced search
2. Click on **View contact as related contact**
3. Select **Benefit Specialist** as relationship type
4. Go to tab cases and select **Housing Support** as case type

Expected results
-----------------------

I expected to see only Contact B as that one has a relationship on the case **Housing Support**.

Actual results
-------------------

I see both Contact B and Contact C.

Comments
----------------------------------------

See also the discussion of a similar related topic: https://lab.civicrm.org/dev/report/-/issues/53

### CODE REVIEW

**Comment by jaapjansma** [comment]
I have fixed the failing test. 
And I have added a test for this specific use case.

File: tests/phpunit/CRM/Case/BAO/QueryTest.php [hunk 1]
```diff
@@ -116,4 +116,62 @@ public function testFindCasesQuery() {
...
+   *
+   * @throws \Exception
+   */
+  public function testAdvancedSearchWithDisplayRelationshipsAndCaseType() {
+    // Preperation
+    $benefitRelationshipTypeId = civicrm_api3('RelationshipType', 'getvalue', ['return' => 'id', 'name_a_b' => 'Benefits Specialist is']);
+    $clientContactID = $this->individualCreate(['first_name' => 'John', 'last_name' => 'Smith']);
+    $benefitSpecialist1 = $this->individualCreate(['Individual', 'first_name' => 'Alexa', 'last_name' => 'Clarke']);
+    $benefitSpecialist2 = $this->individualCreate(['Individual', 'first_name' => 'Sandra', 'last_name' => 'Johnson']);
+    $housingSupportCase = $this->createCase($clientContactID, NULL, ['case_type_id' => 'housing_support', 'case_type_id' => 1]);
```
Comment (demeritcowboy) [inline tests/phpunit/CRM/Case/BAO/QueryTest.php:148]: I think this is a typo: case_type_id used twice.

File: CRM/Contact/BAO/Query.php [hunk 2]
```diff
@@ -5837,10 +5837,11 @@ public function filterRelatedContacts(&$from, &$where, &$having) {
       else {
         $from .= $qcache['from'];
       }
-      $where = $qcache['where'];
+      $where .= $qcache['where'];
```
Comment (demeritcowboy) [inline CRM/Contact/BAO/Query.php:5840]: I kind of remember this code area from the earlier PR but I don't remember if there was ever a time when $where was passed in to the function and is blank or didn't have "WHERE" in it already. I wonder if it's worth adding a guard against that, e.g. if $where does not contain "WHERE", then `$where = "WHERE (1) " . $qcache['where']`. Or something like that - just an example.

File: tests/phpunit/CRM/Case/BAO/QueryTest.php (hunk 1 above)
Comment (jaapjansma) [reply tests/phpunit/CRM/Case/BAO/QueryTest.php:148]: Yes indeed is a typo

File: CRM/Contact/BAO/Query.php (hunk 2 above)
Comment (jaapjansma) [reply CRM/Contact/BAO/Query.php:5843]: I have added a check whether $where is empty.

**Comment by demeritcowboy** [comment]
In case the jenkins log expires:

```
CRM_Contact_SelectorTest::testSelectorQuery with data set #11 (array('Test display relationships', 'CRM_Contact_Selector', array(), array('1_b_a'), null, array(), 'advanced', 0, null, false, array('SELECT contact_a.id as contac...egion`', 'AND displayRelType.relationsh...d = 0)')))
different sql
Failed asserting that two strings are equal.
--- Expected
+++ Actual
@@ @@
-'AND displayRelType.relationship_type_id = 1 AND displayRelType.is_active = 1 AND ( 1 ) AND (contact_a.is_deleted = 0)'
+'WHERE 1 AND displayRelType.relationship_type_id = 1 AND displayRelType.is_active = 1 AND ( 1 ) AND (contact_a.is_deleted = 0)'

/home/jenkins/bknix-dfl/build/core-20002-h980/web/sites/all/modules/civicrm/tests/phpunit/CRM/Contact/SelectorTest.php:66
```

**Comment by jaapjansma** [comment]
Thanks for storing the jenkins log. I have fixed the failing test.

**Comment by demeritcowboy** [comment]
Thanks @jaapjansma I think this is ok to merge.

* General standards
    * [`r-explain`] __PASS__
    * [`r-user`] __PASS__
    * [`r-doc`] __PASS__
    * [`r-run`] __PASS__
      * Can reproduce and this fixes it.
* Developer standards
    * [`r-tech`] __PASS__
      * I'm going to rely on my [earlier look](https://github.com/civicrm/civicrm-core/pull/18779) at this code area where I said:
        > There's already code that does str_replace and if strpos, so it isn't doing something new. ... This code is only ever called when _displayRelationshipType is set to something, and I'm pretty sure, at least in core, that this only happens from the advanced search form (via CRM_Contact_Selector). It's technically possible to call CRM_Contact_BAO_Query::__construct in a way you could end up here, but just browsing a few I didn't see any. I didn't see a way to trigger this via api.
    * [`r-code`] __PASS__
    * [`r-maint`] __PASS__
    * [`r-test`] __PASS__

**Comment by eileenmcnaughton** [comment]
merging based on @demeritcowboy review

**Comment by eileenmcnaughton** [comment]
Just noting this appears to have caused a regression - https://github.com/civicrm/civicrm-core/pull/20997

**Comment by eileenmcnaughton** [comment]
@jaapjansma we made the call to revert this for now in order to fix the regression  & get the releases out. @jmcclelland added a test for the regression so we can be confident it won't regress again. Note your test was not reverted - just marked incomplete for now.

I will re-open in gitlab

**Comment by eileenmcnaughton** [comment]
Actually the original issue had a whole lot of stuff in it not just this so I opened https://lab.civicrm.org/dev/report/-/issues/72 to track

**Pull request merged**